File: Data Structures/2. Linked Lists/06. Print in Reverse/Solution.py
# https://www.hackerrank.com/challenges/print-the-elements-of-a-linked-list-in-reverse/problem

def reversePrint(head):
    if head:
        stack = []
        while head:
            stack.append(head.data)
            head = head.next
        while stack:
            print(stack.pop())
            

def reversePrintRecursive(head):
    if head:
       reversePrint(head.next)
       print(head.data)

File: Data Structures/2. Linked Lists/06. Print in Reverse/Solution.java
// https://www.hackerrank.com/challenges/print-the-elements-of-a-linked-list-in-reverse/problem

static void reversePrint(SinglyLinkedListNode head) {
    Stack stk = new Stack();
    while(head != null){
        stk.push(head.data);
        head = head.next;
    }
    while(!stk.empty()){
        System.out.println(stk.pop());
    }
}

static void reversePrintRecursive(SinglyLinkedListNode head) {
    if(head != null){
        reversePrint(head.next);
        System.out.println(head.data);
    }
  }